core/src/main/java/org/apache/brooklyn/core/catalog/internal/CatalogUpgradeScanner.java

```diff
@@ -0,0 +1,106 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.brooklyn.core.catalog.internal;
+
+import com.google.common.base.Predicate;
+import org.apache.brooklyn.api.typereg.ManagedBundle;
+import org.apache.brooklyn.api.typereg.OsgiBundleWithUrl;
+import org.apache.brooklyn.api.typereg.RegisteredType;
+import org.apache.brooklyn.core.mgmt.ha.OsgiManager;
+import org.apache.brooklyn.core.mgmt.internal.ManagementContextInternal;
+import org.apache.brooklyn.core.typereg.BundleUpgradeParser.CatalogUpgrades;
+import org.apache.brooklyn.util.guava.Maybe;
+import org.osgi.framework.Bundle;
+import org.osgi.framework.BundleContext;
+
+import java.util.Collection;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+
+/**
+ * Scans managed bundles and other jar bundles to find upgrades for installed bundles.
+ */
+class CatalogUpgradeScanner {
+
+    private final ManagementContextInternal managementContext;
+    private final BiFunction<Bundle, RegisteredTypesSupplier, CatalogUpgrades> bundleUpgradeParser;
+    private final Function<OsgiBundleWithUrl, Predicate<? super RegisteredType>> managedBundlePredicateSupplier;
+    private final Function<String, Predicate<? super RegisteredType>> unmanagedBundlePredicateSupplier;
+
+    CatalogUpgradeScanner(
+            final ManagementContextInternal managementContext,
+            final BiFunction<Bundle, RegisteredTypesSupplier, CatalogUpgrades> bundleUpgradeParser,
+            final Function<OsgiBundleWithUrl, Predicate<? super RegisteredType>> managedBundlePredicateSupplier,
+            final Function<String, Predicate<? super RegisteredType>> unmanagedBundlePredicateSupplier
+    ) {
+        this.managementContext = managementContext;
+        this.bundleUpgradeParser = bundleUpgradeParser;
```

Comment:
If subsequently we use the field without null checks, then preference for doing:
```
import static com.google.common.base.Preconditions.checkNotNull;
...
this.bundleUpgradeParser = checkNotNull(bundleUpgradeParser, "bundleUpgradeParser");
```

Reply:
JDK now offers `Objects.requireNonNull(..)` that has the same behaviour.